jaq-std/src/lib.rs

```
@@ -280,7 +328,7 @@ where
 }
 
 /// Convert a string into an array of its Unicode codepoints.
-fn explode<V: ValT>(s: &str) -> impl Iterator<Item = ValR<V>> + '_ {
+fn explode<V: ValT>(s: &[u8]) -> impl Iterator<Item = ValR<V>> + '_ {
     // conversion from u32 to isize may fail on 32-bit systems for high values of c
     let conv = |c: char| Ok(isize::try_from(c as u32).map_err(Error::str)?.into());
     s.chars().map(conv)
```

Hmm.. It turns out the `bstr` version of `chars` automatically does a lossy conversion here (presumably to maintain source compatibility with `str::chars`, which is infallible). This means binary data gets turned into replacement characters, so the original data is not recoverable.

My feeling is that for binary correctness, lossy conversions should generally be avoided, so it should either fail on invalid UTF-8 or it should encode the invalid UTF-8 such that it can be recovered again with `implode`.

I think the most obvious way to achieve the latter is by negating the invalid bytes to get integers in the range `-255` to `-128`, since negative numbers are at least traditionally used for errors in other contexts.

---

I think that I agree about mapping invalid bytes to negative numbers. However, it may not be exactly trivial to do this here, because `bstr::Chars` does not give you an easy access to the invalid bytes (those that you want to map to `-128 ..= -255`). However, it should be doable somehow to reconstruct this. Given your experience with UTF-8, would you like to make this change?

---

Heh, I actually did implement this myself the other day, though I didn't push it anywhere.

I'll push my version and you can decide whether to use it or pick out certain bits etc.

---

This is my version, which I think is a bit more concise, and it passes my [random data test](https://github.com/Maxdamantus/jaq/blob/e20e75b8764af44e978e7e5ff24616377d2d1c84/tests/base64-sh#L18):

https://github.com/Maxdamantus/jaq/commit/aac331aa9f03d178b4ad6c98f2c2b8c5b53e4f2a

Compared to your version, it doesn't pre-allocate (`Vec::with_capacity`), but feel free to add that back in. `xs.len() * 4` seems excessive to me however, since it's very rare that there'll be Unicode text that actually uses 4 bytes per character. I'd probably just hint 1 byte per character, because I'd guess most strings would be ASCII anyway.

I think also your `Iterator::size_hint` implementation is unused. I haven't included an implementation of that.

I kind of wanted it to fail on invalid (positive) code points instead of returning the replacement character, but I've left the replacement character behaviour in for now since there's a test case depending on it. It seems weird however that `[1114112] | implode` returns a replacement character but `[11141120000] | implode` fails. Both of those numbers should just be considered invalid (and unhandled) code points imo and should simply fail.

I believe the current behaviour on master is to fail when imploding invalid code points instead of emitting replacement characters (this is my preferred behaviour):
```sh
$ jaq -n '[1114112] | implode'
Error: "cannot use 1114112 as character"
```

---

Oh, your version looks really nice! I'll definitely use some stuff from there. Nice work!

---

I agree about what you write concerning the replacement character.

---

I made a mix of our two versions at https://github.com/01mf02/jaq/pull/284/commits/c682e7010f6e9bb662a06f087d0d65cfd5d044ae. Thanks a lot again for your feedback!